extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -154,150 +130,60 @@ public List<WritableFrameChannel> outputChannels()
-      if (outputRow == null) {
-        outputRow = currentRow;
-        rowsToProcess.add(currentRow);
-      } else if (comparePartitionKeys(outputRow, currentRow, partitionColumnNames)) {
-        // Add current row to the same batch of rows for processing.
-        rowsToProcess.add(currentRow);
-        if (rowsToProcess.size() > maxRowsMaterialized) {
-          // We don't want to materialize more than maxRowsMaterialized rows at any point in time, so process the pending batch.
-          processRowsUpToLastPartition();
+      if (needToProcessBatch()) {
+        runAllOpsOnBatch();
+        try {
+          flushAllRowsAndCols(resultRowAndCols);
+        }
+        catch (IOException e) {
+          throw new RuntimeException(e);
         }
-        ensureMaxRowsInAWindowConstraint(rowsToProcess.size());
-      } else {
-        lastPartitionIndex = rowsToProcess.size() - 1;
-        outputRow = currentRow.copy();
-        return ReturnOrAwait.runAgain();
       }
-      frameCursor.advance();
+      return ReturnOrAwait.runAgain();
+    } else if (inputChannel.isFinished()) {
+      runAllOpsOnBatch();
+      return ReturnOrAwait.returnObject(Unit.instance());
+    } else {
+      return ReturnOrAwait.awaitAll(inputChannels().size());
     }
-    return ReturnOrAwait.runAgain();
   }
 
-  /**
-   * @param listOfRacs Concat this list of {@link RowsAndColumns} to a {@link ConcatRowsAndColumns} to use as a single input for the operators to be run
-   */
-  private void runAllOpsOnMultipleRac(ArrayList<RowsAndColumns> listOfRacs)
+  private void initialiseOperator()
   {
-    Operator op = new Operator()
+    op = new Operator()
     {
       @Nullable
       @Override
       public Closeable goOrContinue(Closeable continuationObject, Receiver receiver)
       {
-        RowsAndColumns rac = new ConcatRowsAndColumns(listOfRacs);
+        RowsAndColumns rac = new ConcatRowsAndColumns(new ArrayList<>(frameRowsAndCols));
+        frameRowsAndCols.clear();
+        numRowsInFrameRowsAndCols = 0;
         ensureMaxRowsInAWindowConstraint(rac.numRows());
         receiver.push(rac);
-        receiver.completed();
-        return null;
+
+        if (inputChannel.isFinished()) {
+          // Only call completed() when the input channel is finished.
+          receiver.completed();
+          return null; // Signal that the operator has completed its work
+        }
+
+        // Return a non-null continuation object to indicate that we want to continue processing.
+        return () -> {};
       }
     };
-    runOperatorsAfterThis(op);
-  }
-
-  /**
-   * @param op Base operator for the operators to be run. Other operators are wrapped under this to run
-   */
-  private void runOperatorsAfterThis(Operator op)
-  {
     for (OperatorFactory of : operatorFactoryList) {
       op = of.wrap(op);
     }
-    Operator.go(op, new Operator.Receiver()
+  }
+
+  private void runAllOpsOnBatch()
+  {
+    op.goOrContinue(null, new Operator.Receiver()
```

Comment:
I originally tried putting `flushAllRowsAndCols(resultRowAndCols)` inside  `runAllOpsOnBatch() -> push()` method, like following:
```java
private void runAllOpsOnBatch()
  {
    op.goOrContinue(null, new Operator.Receiver()
    {
      @Override
      public Operator.Signal push(RowsAndColumns rac)
      {
        resultRowAndCols.add(rac);
        try{
          flushAllRowsAndCols(resultRowAndCols);
        }
        catch (IOException e) {
          throw new RuntimeException(e);
        }
        return Operator.Signal.GO;
      }

      @Override
      public void completed()
      {
        try {
          flushAllRowsAndCols(resultRowAndCols);
        }
        catch (IOException e) {
          throw new RuntimeException(e);
        }
      }
    });
  }
```

But this runs into the following error:
```java
java.lang.RuntimeException: org.apache.druid.java.util.common.ISE: Channel has no capacity
```

When I dug further into this, it seems that the contract of `runIncrementally()` doesn't allow writing multiple frames in a single iteration of `runIncrementally()`, which ends up happening in this case as the `push()` can get called multiple times within a single iteration of `runIncrementally()`.

I also stumbled upon this PR from Gian where he had to handle this in`ScanQueryFrameProcessor` as well: https://github.com/apache/druid/pull/13036

I'm not sure of the historic reason behind this design. But because of these restrictions, I had to move `flushAllRowsAndCols(resultRowAndCols)` into `runIncrementally()`.